File: clox/debug.c
#include <stdio.h>

#include "debug.h"
#include "value.h"

void disassembleChunk(Chunk *chunk, const char *name) {
    printf("== %s ==\n", name);

    for (int offset = 0; offset <chunk->count;) {
        offset = disassembleInstruction(chunk, offset); // this call returns offset of NEXT instruction (since instructions can be of diff. sizes)
    }
}

static int constantInstruction(const char* name, Chunk* chunk, int offset) {
    uint8_t constant = chunk->code[offset + 1];
    printf("%-16s %4d '", name, constant);
    printValue(chunk->constants.values[constant]);
    printf("'\n");

    return offset + 2;
}

static int simpleInstruction(const char* name, int offset) {
    printf("%s\n", name);
    return offset + 1;
}

int disassembleInstruction(Chunk *chunk, int offset) {
    printf("%04d ", offset); // where in chunk this instruction is
    // note: & = start of format specifier :: 0 = num padded with zeroes (not spaces) if shorter than specified width :: 4 = min width of output field :: d = expect integer, format as decimal num'
 if (offset > 0 && chunk->lines[offset] == chunk->lines[offset - 1]) {
     printf("  | ");
 } else {
     printf("%4d ", chunk->lines[offset]);
 }
    uint8_t instruction = chunk->code[offset]; // read single byte (opcode)
    switch (instruction) {
        case OP_RETURN:
            return simpleInstruction("OP_RETURN", offset);
        case OP_CONSTANT:
            return constantInstruction("OP_CONSTANT", chunk, offset);
        default:
            printf("Unknown opcode %d\n", instruction);
            return offset + 1;
    }
}

File: rlox/my-bytecode-interpreter/src/debug.rs
use crate::{Chunk, OpCode, Value};

pub fn disassemble(chunk: &Chunk, name: &str) {
    println!("== {} ==\n", name);

    let mut offset = 0;
    while offset < chunk.code.len() {
        offset = disassemble_instruction(chunk, offset);
    }
}

pub fn disassemble_instruction(chunk: &Chunk, offset: usize) -> usize {
    print!("{:04} ", offset);

    if offset > 0 && chunk.lines[offset] == chunk.lines[offset - 1] {
        print!("  | ")
    } else {
        print!("{:4} ", chunk.lines[offset]);
    }

    let instruction = OpCode::from(chunk.code[offset]);

    match instruction {
        OpCode::Return => simple_instruction(&"OP_RETURN", offset),
        OpCode::Constant => constant_instruction(&"OP_CONSTANT", chunk, offset),
        OpCode::Negate => simple_instruction("OP_NEGATE", offset),
        OpCode::Add => simple_instruction("OP_ADD", offset),
        OpCode::Subtract => simple_instruction("OP_SUBTRACT", offset),
        OpCode::Multiply => simple_instruction("OP_MULTIPLY", offset),
        OpCode::Divide => simple_instruction("OP_DIVIDE", offset),
    }
}

fn simple_instruction(name: &str, offset: usize) -> usize {
    println!("{}", name);
    offset + 1
}

fn constant_instruction(name: &str, chunk: &Chunk, offset: usize) -> usize {
    let constant = chunk.code[offset + 1];
    if constant as usize >= chunk.constants.len() {
        panic!(
            "DEBUG: constant idx out of bounds: {} >= {}",
            constant,
            chunk.constants.len()
        );
    }

    print!("{:-16} {:4} '", name, constant);
    print_value(&chunk.constants[constant as usize]);
    println!("'");
    offset + 2
}

pub fn print_value(value: &Value) {
    match value {
        Value::Number(n) => print!("{}", n),
    }
}
